File: elixir/wordy/lib/wordy.ex
defmodule Wordy do
  @doc """
  Calculate the math problem in the sentence.
  """
  @spec answer(String.t()) :: integer
  def answer(question) do
    question
    |> parse_question()
    |> validate_prefix_format()
    |> validate_suffix_question()
    |> parse_operators()
    |> validate_operators()
    |> evaluate()
    |> format()
  end

  defp parse_question(question) do
    question
    |> String.replace("plus", "+")
    |> String.replace("minus", "-")
    |> String.replace("multiplied by", "*")
    |> String.replace("divided by", "/")
    |> String.split(" ", trim: true)
    |> fn tokens -> {:ok, tokens} end.()
  end

  defp validate_prefix_format(:error), do: :error
  defp validate_prefix_format({:ok, tokens}) do
    case tokens do
      ["What", "is" | tail] -> {:ok, tail}
      _ -> :error
    end
  end

  defp validate_suffix_question(:error), do: :error
  defp validate_suffix_question({:ok, tokens}) do
    last = tokens |> List.last()
    case last |> String.reverse() do
      "?" <> reversed ->
        rest = tokens
        |> Enum.reverse()
        |> tl()
        [reversed |> String.reverse() | rest]
        |> Enum.reverse()
        |> fn parsed -> {:ok, parsed} end.()
      _ -> :error
    end
  end

  defp parse_operators(:error), do: :error
  defp parse_operators({:ok, tokens}) do
    tokens
    |> Enum.map(fn
      "+" -> :add
      "-" -> :subtract
      "*" -> :multiply
      "/" -> :divide
      token -> case Integer.parse(token) do
        {number, ""} -> number
        _ -> :error
      end
    end)
    |> Enum.reduce({:ok, []}, fn
      _, :error -> :error
      :error, _ -> :error
      token, {:ok, tokens} -> {:ok, [token | tokens]}
    end)
    |> fn
      {:ok, tokens} -> {:ok, tokens |> Enum.reverse() }
      :error -> :error
    end.()
  end

  defp operator?(:add), do: true
  defp operator?(:minus), do: true
  defp operator?(:multiply), do: true
  defp operator?(:divide), do: true
  defp operator?(_), do: false

  defp validate_operators(:error), do: :error
  defp validate_operators({:ok, tokens}) do
    [head | tail] = tokens
    tail
    |> Enum.reduce({:ok, head}, fn
      _, :error -> :error
      token, {:ok, value} ->
        cond do
          operator?(token) and operator?(value) -> :error
          true -> {:ok, token}
        end
    end)
    |> fn
      :error -> :error
      _ -> {:ok, tokens}
    end.()
  end


  defp evaluate(:error), do: :error
  defp evaluate({:ok, tokens}) do
    case tokens do
      [var_one, :add, var_two | tail] -> evaluate({:ok, [var_one + var_two | tail]})
      [var_one, :subtract, var_two | tail] -> evaluate({:ok, [var_one - var_two | tail]})
      [var_one, :multiply, var_two | tail] -> evaluate({:ok, [var_one * var_two | tail]})
      [var_one, :divide, var_two | tail] -> evaluate({:ok, [div(var_one,var_two) | tail]})
      [result] -> {:ok, result}
      _ -> :error
    end
  end

  defp format(:error), do: raise ArgumentError
  defp format({:ok, result}), do: result
end

File: c/wordy/wordy.c
#include "wordy.h"
#include <ctype.h>
#include <stdbool.h>
#include <string.h>

const char *prefix = "What is ";
const int prefix_length = sizeof("What is ") - 1;
const char *plus = " plus ";
const int plus_length = sizeof(" plus ") - 1;
const char *minus = " minus ";
const int minus_length = sizeof(" minus ") - 1;
const char *multiply = " multiplied by ";
const int multiply_length = sizeof(" multiplied by ") - 1;
const char *divide = " divided by ";
const int divide_length = sizeof(" divided by ") - 1;

typedef enum token_type_t {
  Number,
  Plus,
  Minus,
  Multiply,
  Divide,
} token_type_t;

typedef struct token_t {
  token_type_t type;
  int number;
} token_t;

static token_t read_token(const char *question, int *read_index) {
  if (strncmp(&question[*read_index], plus, plus_length) == 0) {
    *read_index += plus_length;
    return (token_t){.type = Plus};
  } else if (strncmp(&question[*read_index], minus, minus_length) == 0) {
    *read_index += minus_length;
    return (token_t){.type = Minus};
  } else if (strncmp(&question[*read_index], multiply, multiply_length) == 0) {
    *read_index += multiply_length;
    return (token_t){.type = Multiply};
  } else if (strncmp(&question[*read_index], divide, divide_length) == 0) {
    *read_index += divide_length;
    return (token_t){.type = Divide};
  }
  bool is_negative = false;
  if (question[*read_index] == '-') {
    is_negative = true;
    ++*read_index;
  }
  int start = *read_index;
  while (isdigit(question[*read_index])) {
    ++*read_index;
  }
  int d = 0;
  for (int i = start; i < *read_index; ++i) {
    d = d * 10 + (question[i] - '0');
  }
  return (token_t){
      .type = Number,
      .number = is_negative ? -d : d,
  };
}

bool answer(const char *question, int *result) {
  const int question_length = strlen(question);
  const int prefix_length = strlen(prefix);
  if (strncmp(question, prefix, prefix_length) != 0) {
    return false;
  }
  int read_index = prefix_length;
  token_t t = read_token(question, &read_index);
  if (t.type != Number) {
    return false;
  }
  *result = t.number;
  while (read_index < question_length && question[read_index] != '?') {
    t = read_token(question, &read_index);
    switch (t.type) {
    case Number:
      return false;
    case Plus:
      t = read_token(question, &read_index);
      if (t.type != Number) {
        return false;
      }
      *result += t.number;
      break;
    case Minus:
      t = read_token(question, &read_index);
      if (t.type != Number) {
        return false;
      }
      *result -= t.number;
      break;
    case Multiply:
      t = read_token(question, &read_index);
      if (t.type != Number) {
        return false;
      }
      *result *= t.number;
      break;
    case Divide:
      t = read_token(question, &read_index);
      if (t.type != Number) {
        return false;
      }
      *result /= t.number;
      break;
    }
  }
  return true;
}
